lte/gateway/python/magma/pipelined/app/apn.py

```diff
@@ -0,0 +1,284 @@
+            
+            Args:
+                generator: function which responsible for add / delete flow mod
+                    generation
+                matcher: MagmaMatcher instance which specifies traffic flow 
+                    direction
+                apn_tagging_actions: actions to be applied to packets at 
+                    ScratchTable level
+
+        """
+        return [
+            generator(
+                self._datapath, 
+                self.tbl_num,
+                matcher,
+                None,
+                resubmit_table=self.apn_tagging_scratch,
+                priority=flows.DEFAULT_PRIORITY),
+            generator(
+                self._datapath, 
+                self.apn_tagging_scratch,
+                matcher,
+                apn_tagging_actions,
+                resubmit_table=self.next_table,
+                priority=flows.DEFAULT_PRIORITY)]
+    
+    def delete_apn_flow_for_ue(self, imsi, ue_ip_addr, apn):
+        """ Delete flow been created in scope of add_apn_flow_for_ue.
+
+        Args:
+            imsi: user's IMSI
+            ue_ip_addr: ip addr allocated for the UE in scope of connection to 
+                specific APN
+            apn: APN UE is connected to with specified IP addr
+        """
+        #TODO(119vik): same IP is reused for several bearers connected to the 
+        # same APN - take care about duplications
+        parser = self._datapath.ofproto_parser
+        
+        encoded_apn = encode_apn(apn)
+        apn_tagging_actions = self._get_apn_tagging_actions(parser, encoded_apn)
+        flow_dels = []
+
+        # Prepare matchers for both UpLink and DownLink traffic
+        outbound_match = MagmaMatch(
+            direction=Direction.OUT, 
+            ipv4_src=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        inbound_match = MagmaMatch(
+            direction=Direction.IN, 
+            ipv4_dst=ue_ip_addr, 
+            eth_type=ether.ETH_TYPE_IP)
+        
+        # Generate OF Add Flow messages for Uplink and DownLink traffic
+        flow_dels += self._get_flows_for_traffic_direction(
+            get_del_resubmit_next_service_flow_msg,
+            outbound_match, 
+            apn_tagging_actions
+        )
+        flow_dels += self._get_flows_for_traffic_direction(
+            get_del_resubmit_next_service_flow_msg,
+            inbound_match, 
+            apn_tagging_actions
+        )
+        logging.info("Flows to delete {}, datapath {}".format(flow_dels, 
+                                                              self._datapath))
+        
+        # Push messages to OVS through async message hub
+        chan = self._msg_hub.send(flow_dels, self._datapath)
+
+        # Wait and process response
+        return self._wait_for_flow_responses(imsi, flow_dels, chan)  
+
+
+# TODO(119vik): migrate following to the openflow.flows module
+def get_del_resubmit_next_service_flow_msg(datapath, table, match, actions=None,
```

Comment:
I'm a bit confused what this does. The `get_delete_flow_msg` deletes the flow,  are you looking for something like add_resubmit_next_service_flow? You can add an action so it also copied to the scratch

Reply:
sorry for unclear comments.  During bearer creation i'm creating resubmit flow using functions from flows module - here i've function which helps me construct "delete flow" message for resubmit flow which was created earlier. I'm going to use this function in scope of bearer destroy sequence.

Reply:
imo this function is unnecessary - I would just expand it at its callsite